cmd/ketch/builder_list.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+)
+
+const builderListHelp = `
+List CNCF registered builders, along with any additional builders defined by the user in config.toml (default path: $HOME/.kafka)
+`
+
+type builderListEntry struct {
+	Vendor      string
+	Image       string
+	Description string
+}
+
+var builderList = []builderListEntry{
+	{
+		Vendor:      "Google",
+		Image:       "gcr.io/buildpacks/builder:v1",
+		Description: "GCP Builder for all runtimes",
+	},
+	{
+		Vendor:      "Heroku",
+		Image:       "heroku/buildpacks:18",
+		Description: "heroku-18 base image with buildpacks for Ruby, Java, Node.js, Python, Golang, & PHP",
+	},
+	{
+		Vendor:      "Heroku",
+		Image:       "heroku/buildpacks:20",
+		Description: "heroku-20 base image with buildpacks for Ruby, Java, Node.js, Python, Golang, & PHP",
+	},
+	{
+		Vendor:      "Paketo Buildpacks",
+		Image:       "paketobuildpacks/builder:base",
+		Description: "Small base image with buildpacks for Java, Node.js, Golang, & .NET Core",
+	},
+	{
+		Vendor:      "Paketo Buildpacks",
+		Image:       "paketobuildpacks/builder:full",
+		Description: "Larger base image with buildpacks for Java, Node.js, Golang, .NET Core, & PHP",
+	},
+	{
+		Vendor:      "Paketo Buildpacks",
+		Image:       "paketobuildpacks/builder:tiny",
+		Description: "Tiny base image (bionic build image, distroless run image) with buildpacks for Golang",
+	},
+}
+
+func newBuilderListCmd(cfg config, out io.Writer) *cobra.Command {
```

Comment:
If this function took `ketchConfig configuration.KetchConfig` instead of `cfg config`, it would ease testing and simplify `config` interface.

```golang
type config interface {
	Client() client.Client
	Storage() templates.Client
      
       // ...

        // no need to add this function to the interface
	// GetKetchConfigObject returns the unmarshalled contents of the config.toml file
	// GetKetchConfigObject() configuration.KetchConfig
}

func newBuilderListCmd(ketchConfig configuration.KetchConfig, out io.Writer) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: builderListHelp,
		Long:  builderListHelp,
		Args:  cobra.NoArgs,
		Run: func(cmd *cobra.Command, args []string) {
			writeBuilders(ketchConfig, out)
		},
	}
	return cmd
}

func getUserBuilders(ketchConfig configuration.KetchConfig) []builderListEntry {
	var userBuilders []builderListEntry
	for _, builder := range ketchConfig.AdditionalBuilders {
		entry := builderListEntry{
			Vendor:      builder.Vendor,
			Image:       builder.Image,
			Description: builder.Description,
		}
		userBuilders = append(userBuilders, entry)
	}

	return userBuilders
}

func writeBuilders(ketchConfig configuration.KetchConfig, out io.Writer) {
	tw := tabwriter.NewWriter(out, 10, 10, 5, ' ', 0)

	builderList = append(builderList, getUserBuilders(ketchConfig)...)

	fmt.Fprintln(tw, "VENDOR\tIMAGE\tDESCRIPTION")
	for _, builder := range builderList {
		fmt.Fprintf(tw, "%s:\t%s\t%s\t\n", builder.Vendor, builder.Image, builder.Description)
	}

	tw.Flush()
}

```

Reply:
That cleaned up a tremendous amount of code! Thank you for the suggestion.

The reason I added the method in the first place was due to not being able to access the data directly due to the interface being used. I hadn't thought of converting back to the actual type at the point where I actually needed the data.